src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs

```diff
@@ -0,0 +1,280 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        private sealed partial class ShuffleIterator<TSource>
+        {
+            public override TSource[] ToArray()
+            {
+                TSource[] array = _source.ToArray();
+                Random.Shared.Shuffle(array);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                List<TSource> list = _source.ToList();
+                Random.Shared.Shuffle(CollectionsMarshal.AsSpan(list));
+                return list;
+            }
+
+            public override int GetCount(bool onlyIfCheap) =>
+                !onlyIfCheap ? _source.Count() :
+                TryGetNonEnumeratedCount(_source, out int count) ? count :
+                -1;
+
+            public override TSource? TryGetFirst(out bool found) =>
+                TryGetElementAt(0, out found);
+
+            public override TSource? TryGetLast(out bool found) =>
+                TryGetElementAt(0, out found);
```

Comment:
@stephentoub not sure if TryGetLast is impl correctly here?

Reply:
It is. This is a shuffle, so the index doesn't matter; last is the same as first.